src/A2A/BaseKindDiscriminatorConverter.cs

```diff
@@ -0,0 +1,87 @@
+using System.Globalization;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Text.Json.Serialization.Metadata;
+
+namespace A2A;
+
+/// <summary>
+/// Base class for JSON converters that use a "kind" discriminator to deserialize to different derived types.
+/// </summary>
+/// <typeparam name="TBase">The base type being converted.</typeparam>
+/// <typeparam name="TKind">The enum type used as the discriminator.</typeparam>
+internal abstract class BaseKindDiscriminatorConverter<TBase, TKind> : JsonConverter<TBase>
+    where TBase : class
+    where TKind : struct, Enum
+{
+    internal const string DiscriminatorPropertyName = "kind";
+
+    /// <summary>
+    /// Gets the mapping from kind enum values to their corresponding concrete types.
+    /// </summary>
+    protected abstract Type?[] TypeMapping { get; }
+
+    /// <summary>
+    /// Gets the entity name used in error messages (e.g., "part", "file content", "event").
+    /// </summary>
+    protected abstract string DisplayName { get; }
+
+    /// <summary>
+    /// Deserializes the kind enum value from the JSON property using the appropriate JsonTypeInfo.
+    /// </summary>
+    /// <param name="kindProp">The JSON property containing the kind value.</param>
+    /// <returns>The deserialized enum value.</returns>
+    protected abstract TKind DeserializeKind(JsonElement kindProp);
+
+    public override TBase Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+    {
+        using var document = JsonDocument.ParseValue(ref reader);
+        var root = document.RootElement;
+
+        if (!root.TryGetProperty(DiscriminatorPropertyName, out var kindProp) || kindProp.ValueKind is not JsonValueKind.String)
+        {
+            throw new A2AException($"Missing required '{DiscriminatorPropertyName}' discriminator for {typeof(TBase).Name}.", A2AErrorCode.InvalidRequest);
+        }
+
+        TBase? obj = null;
+        Exception? deserializationException = null;
+        try
+        {
+            var kindValue = DeserializeKind(kindProp);
+            var kindToTypeMapping = TypeMapping;
+            var kindIndex = Convert.ToInt32(kindValue, CultureInfo.InvariantCulture);
+
+            if (kindIndex < 0 || kindIndex >= kindToTypeMapping.Length || kindToTypeMapping[kindIndex] is not Type targetType)
```

Comment:
I believe `kindToTypeMapping[kindIndex] is not Type targetType` is dirrefent failure mode. Error message is misleading as this is known kind but mismatched to other object. This is common copy/paste mistake in by hand crafted requests and therefore deserve its specific error message.